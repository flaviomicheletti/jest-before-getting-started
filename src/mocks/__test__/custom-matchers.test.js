//
// https://jestjs.io/docs/en/next/mock-functions#custom-matchers
//

const mockFunc = jest
    .fn()
    .mockReturnValue('default')
    .mockImplementation(scalar => 42 + scalar)
    .mockName('add42');

const arg1 = 100;

describe("Custom Matchers", () => {

    test('The mock function was called at least once', () => {
        mockFunc()
        expect(mockFunc).toHaveBeenCalled();
    });

    test('The mock function was called at least once with the specified args', () => {

        mockFunc(arg1)
        expect(mockFunc).toHaveBeenCalledWith(arg1);
    });

    test('All calls and the name of the mock is written as a snapshot', () => {
        expect(mockFunc).toMatchSnapshot();
    });

    test('The mock function was called at least once', () => {
        expect(mockFunc.mock.calls.length).toBeGreaterThan(0);
    });

    test('The mock function was called at least once with the specified args', () => {
        expect(mockFunc.mock.calls).toContainEqual([arg1]);
    });

    test('The last call to the mock function was called with the specified args', () => {
        expect(mockFunc.mock.calls[mockFunc.mock.calls.length - 1]).toEqual([arg1]);
    });

    test('The first arg of the last call to the mock function was `42', () => {
        // (note that there is no sugar helper for this specific of an assertion)        
        expect(mockFunc.mock.calls[mockFunc.mock.calls.length - 1][0]).toBe(100);
    });

    test('A snapshot will check that a mock was invoked the same number of times,', () => {
        // ...in the same order, with the same arguments.
        // It will also assert on the name        
        expect(mockFunc.mock.calls).toEqual([[], [100]]);
        expect(mockFunc.getMockName()).toBe('add42');
    });


});