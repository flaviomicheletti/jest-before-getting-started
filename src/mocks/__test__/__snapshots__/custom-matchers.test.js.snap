// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Matchers All calls and the name of the mock is written as a snapshot 1`] = `
[MockFunction add42] {
  "calls": Array [
    Array [],
    Array [
      100,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": NaN,
    },
    Object {
      "type": "return",
      "value": 142,
    },
  ],
}
`;
